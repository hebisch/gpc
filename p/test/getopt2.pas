{ Test of the GetOpt functions. (Very similar to GetOptDemo.) }

program GetOptTest;

uses GPC;

var
  ch, LongOptFlag : Char;
  i, LongIndex : Integer;

const
  LongOptions : array [1 .. 4] of OptionType =
    (('abc',    NoArgument,       nil,          'x'),
     ('noarg',  NoArgument,       @LongOptFlag, #2),
     ('reqarg', RequiredArgument, @LongOptFlag, #3),
     ('optarg', OptionalArgument, @LongOptFlag, #4));

begin
  LongIndex := - 1;
  GetOptErrorFlag := True;
  repeat
    ch := GetOptLong ('nr:o::-', LongOptions, LongIndex, True);
    case ch of
      EndOfOptions  : Break;
      NoOption      : Write ('no-option argument');
      UnknownOption : if UnknownOptionCharacter = UnknownLongOption
                        then Write ('(incorrect long option)')
                        else Write ('unknown option `', UnknownOptionCharacter, '''');
      LongOption    : with LongOptions [LongIndex] do
                        begin
                          Write ('long option `', CString2String (OptionName), '''');
                          if Ord (LongOptFlag) <> LongIndex then
                            Write (' <internal error> ')
                        end;
      else            Write ('option `', ch, '''')
    end;
    if HasOptionArgument
      then WriteLn (' with argument `', OptionArgument, '''')
      else WriteLn
  until False;
  if (FirstNonOption < 1) or (FirstNonOption > ParamCount + 1) then
    begin
      WriteLn (StdErr, 'Internal error with FirstNonOption.');
      Halt (2)
    end;
  if FirstNonOption <= ParamCount then WriteLn ('Remaining arguments:');
  for i := FirstNonOption to ParamCount do WriteLn (ParamStr (i))
end.
