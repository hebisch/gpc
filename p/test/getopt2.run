#!/bin/sh
#
if [ -r "$A_OUT" ]; then ./"$A_OUT" --optarg -r foo -x -o -nobar --noarg --reqarg foo --optarg=bar --abc baz qux; else echo failed; fi
