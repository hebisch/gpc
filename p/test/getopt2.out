long option `optarg'
option `r' with argument `foo'
option `x'
option `o'
option `n'
option `o' with argument `bar'
long option `noarg'
long option `reqarg' with argument `foo'
long option `optarg' with argument `bar'
option `x'
Remaining arguments:
baz
qux
